aten/src/ATen/core/PhiloxRNGEngine.h

```diff
@@ -192,12 +190,49 @@ class philox_engine {
     ret[3] = lo0;
     return ret;
   }
+
+  C10_HOST_DEVICE constexpr float uint32_to_uniform_float(uint32_t value) {
+      // maximum value such that `MAX_INT * scale < 1.0` (with float rounding)
+      constexpr float scale = 4.6566127342e-10;
+      return static_cast<float>(value & 0x7FFFFFFF) * scale;
+  }
+
+
+
+  C10_HOST_DEVICE inline detail::UINT4 rand(detail::UINT4& counter, detail::UINT2& key, uint32_t n_rounds) {
+    for (uint32_t round = 0; round < (n_rounds - 1); round++) {
```

Comment:
for cuda having nrounds be a compile time constant is very important to unroll the loop (without it perf really tanks), I don't know what perf effect there is on cpu

Reply:
Generally, most guidelines are around not doing this by hand - modern C++ compilers are very good at figuring out when and how to unroll w/r/t branch prediction. Since this will always be a fixed size loop, with the fixed size most likely hardcoded upstream, as opposed to being a product of some I/O - I think leaving it as a loop gives the compiler the most flexibility in how/when to unroll it.